pybmix/core/pybmixcpp/python_hierarchy.cc

```diff
@@ -0,0 +1,233 @@
+
+#include <google/protobuf/stubs/casts.h>
+#include <pybind11/embed.h>
+#include <pybind11/pybind11.h>
+
+#include <Eigen/Dense>
+#include <random>
+#include <sstream>
+#include <stan/math/prim/prob.hpp>
+#include <string>
+#include <vector>
+
+#include "algorithm_state.pb.h"
+#include "hierarchy_prior.pb.h"
+#include "ls_state.pb.h"
+#include "pybmixcpp/bayesmix/src/utils/rng.h"
+
+namespace py = pybind11;
+using namespace py::literals;
+
+extern py::module_ numpy;
+extern py::module_ fun;
+extern py::module_ numpy_random;
+extern py::object py_engine;
+extern py::object py_gen;
+
+void synchronize_cpp_to_py_state(const std::mt19937 &cpp_gen,
+                                 py::object &py_gen);
+
+void synchronize_py_to_cpp_state(std::mt19937 &cpp_gen,
+                                 const py::object &py_gen);
+
+py::list vector_to_list(const std::vector<double> &x);
+
+std::vector<double> list_to_vector(py::list &x);
+
+py::list eigen_to_list(const Eigen::RowVectorXd &x);
+
+//! PYTHON
+double PythonHierarchy::like_lpdf(const Eigen::RowVectorXd &datum) const {
+    py::list state_py = vector_to_list(state.generic_state);
+    py::list datum_py = eigen_to_list(datum);
+    double result = fun.attr("like_lpdf")(datum_py, state_py).cast<double>();
+    return result;
+}
+
+//! PYTHON
+double PythonHierarchy::marg_lpdf(const Python::Hyperparams &params,
+                                  const Eigen::RowVectorXd &datum) const {
+    py::list params_py = vector_to_list(params.generic_hypers);
+    py::list datum_py = eigen_to_list(datum);
+    double result = fun.attr("marg_lpdf")(datum_py, params_py).cast<double>();
+    return result;
+}
+
+//! PYTHON
+void PythonHierarchy::initialize_state() {
+    py::list hypers_py = vector_to_list(hypers->generic_hypers);
+    py::list state_py = fun.attr("initialize_state")(hypers_py);
+    state.generic_state = list_to_vector(state_py);
+}
+
+//! C++
+void PythonHierarchy::initialize_hypers() {
+    if (prior->has_values()) {
+        // Set values
+        hypers->generic_hypers.clear();
+        int size = prior->values().size();
+        for (int i = 0; i < size; ++i) {
+            hypers->generic_hypers.push_back((prior->values().data())[i]);
+        }
+    }
+}
+
+//! PYTHON
+void PythonHierarchy::update_hypers(
+        const std::vector <bayesmix::AlgorithmState::ClusterState> &states) {
+    auto &rng = bayesmix::Rng::Instance().get();
+    if (prior->has_values()) return;
+}
+
+//! PYTHON
+Python::State PythonHierarchy::draw(const Python::Hyperparams &params) {
+  Python::State out;
+  synchronize_cpp_to_py_state(bayesmix::Rng::Instance().get(), py_gen);
+  py::list params_py = vector_to_list(params.generic_hypers);
+  py::list state_py = vector_to_list(state.generic_state);
+  py::list draw_py = fun.attr("draw")(state_py,params_py,py_gen);
+  out.generic_state = list_to_vector(draw_py);
+  synchronize_py_to_cpp_state(bayesmix::Rng::Instance().get(), py_gen);
+  return out;
```

Comment:
Actually, I would delegate the computation of posterior hypers entirely to the "draw" python function, which should just receive the data allocated to this cluster for the moment

Reply:
Honestly, we didn't understand this comment, can you please explain more in details?

Reply:
Right now, you have two functions: `compute_posterior_hypers` and `draw` in `fun.py`

In practice, it's hardly the case that users will want to implement conjugate hierarchies in python. Therefore, I suggest working directly as if the hierarchy is non conjugate, and have the users implement a `draw_prior` and a `sample_full_conditional` pythonfunction. The second one could also call the `draw_prior` (as in this case) but this logic would be within the python code, and therefore be handled only by the user